enhancements/SEP75-automatic-workload-migration.md

```diff
@@ -0,0 +1,643 @@
+      currentBatch: 5
+      totalBatches: 10
+```
+
+**Migration completed successfully:**
+```yaml
+status:
+  workloadMigration:
+    state: Completed
+    targetRevision: default-v1-26-0
+    totalWorkloads: 47
+    migratedWorkloads: 47
+    failedWorkloads: 0
+    startTime: "2025-10-21T10:30:00Z"
+    completionTime: "2025-10-21T10:45:23Z"
+    batched:
+      currentBatch: 10
+      totalBatches: 10
+```
+
+**Migration completed with failures:**
+```yaml
+status:
+  workloadMigration:
+    state: Failed
+    targetRevision: default-v1-26-0
+    totalWorkloads: 47
+    migratedWorkloads: 44
+    failedWorkloads: 3
+    failures:
+      - namespace: production
+        name: legacy-app
+        kind: Deployment
+        reason: "Readiness timeout exceeded after 5m0s"
+        timestamp: "2025-10-21T10:35:12Z"
+      - namespace: staging
+        name: flaky-service
+        kind: Deployment
+        reason: "Deployment failed to reach ready state: 2/3 replicas ready"
+        timestamp: "2025-10-21T10:38:45Z"
+      - namespace: production
+        name: critical-app
+        kind: Deployment
+        reason: "Readiness timeout exceeded after 5m0s"
+        timestamp: "2025-10-21T10:42:30Z"
+    startTime: "2025-10-21T10:30:00Z"
+    completionTime: "2025-10-21T10:45:23Z"
+    batched:
+      currentBatch: 8
+      totalBatches: 10
+```
+
+##### Status Update Frequency
+
+- Status is updated at the start and end of each batch
+- Status counters are updated as each workload completes (success or failure)
+- The `Failures` array in status is limited to 10 most recent entries to prevent unbounded growth
+
+#### WorkloadManager & InUse Detection
+
+The introduction of a `WorkloadManager` is proposed that will handle workload-specific tasks such as label updates and InUse detection. Currently, InUse detection code is spread over the `Istio`/`IstioRevision`/`IstioRevisionTag` controllers. Implementation of this SEP should include a refactoring that moves the code into a common package.
+
+#### Deployment Restart Mechanism
+
+Deployments are restarted using Kubernetes' standard rolling update mechanism:
+
+1. **Label Update**: Pod template labels are updated to reference the new revision, if required - the only exception being when `IstioRevisionTag`s are used, and the referenced `Istio` resource is the one being updated
+2. **Restart Annotation**: If a pod label change was not required, a `kubectl.kubernetes.io/restartedAt` annotation is added to trigger pod replacement
+3. **Health Check**: The operator waits for `deployment.Status.ReadyReplicas == deployment.Spec.Replicas`
+
+### Performance Impact
+
+* **Discovery Overhead**: The operator lists all namespaces and deployments once per migration
+* **Batch Processing**: Migration impact is controlled through configurable batch sizes and delays
+* **Memory Usage**: Minimal additional memory for tracking migration state
+* **Network Traffic**: Standard Kubernetes API calls for object updates
```

Comment:
If the whole migration is performed in a single reconcile loop, I'm more worried about the operator being tied up with the migration for a particular `Istio` resource and unable to respond to updates to that `Istio` resource until the migration is finished which could take awhile given a large amount of workloads and needing to wait until each pod has restarted.